src/lj_arena.c

```diff
@@ -0,0 +1,381 @@
+/* To get the mremap prototype. Must be defined before any system includes. */
+#if defined(__linux__) && !defined(_GNU_SOURCE)
+#define _GNU_SOURCE
+#endif
+
+#include "lj_def.h"
+#include "lj_arena.h"
+#include "lj_gc.h"
+#include "lj_obj.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+/* unmap() any misaligned huge pages */
+#define LJ_ALLOC_FIXUP 1
+
+#if LJ_TARGET_LINUX
+#define LJ_ALLOC_MREMAP 1
+#endif
+
+
+#if LJ_TARGET_POSIX
+#include <sys/mman.h>
+#include <pthread.h>
+
+#define RESERVE_PAGES(h, n)                                                    \
+  mmap(h, n, PROT_WRITE | PROT_READ, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0)
+#define UNRESERVE_PAGES(p, n) munmap(p, n)
+#define COMMIT_PAGES(p, n)
+#define UNCOMMIT_PAGES(p, n) /*madvise(p, n, MADV_DONTNEED)*/
+#define RESERVE_AND_COMMIT_PAGES(n) RESERVE_PAGES(0, n)
+
+static pthread_mutex_t g_lock;
+#define LOCK() pthread_mutex_lock(&g_lock)
+#define UNLOCK() pthread_mutex_unlock(&g_lock)
+
+#elif LJ_TARGET_WINDOWS
+#define COMMITLESS
+
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+
+#define UNRESERVE_PAGES(p, n) VirtualFree(p, 0, MEM_RELEASE)
+#define RESERVE_AND_COMMIT_PAGES(n)                                            \
+  LJ_WIN_VALLOC(0, n, MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE)
+#ifdef COMMITLESS
+#define RESERVE_PAGES(h, n) LJ_WIN_VALLOC(h, n, MEM_RESERVE|MEM_COMMIT, PAGE_READWRITE)
+#define COMMIT_PAGES(p, n)
+#define UNCOMMIT_PAGES(p, n)
+#else
+#define RESERVE_PAGES(h, n) LJ_WIN_VALLOC(h, n, MEM_RESERVE, PAGE_READWRITE)
+#define COMMIT_PAGES(p, n)                                                     \
+  if (!LJ_WIN_VALLOC(p, n, MEM_COMMIT, PAGE_READWRITE))                        \
+  commit_error(p)
+#define UNCOMMIT_PAGES(p, n) VirtualFree(p, n, MEM_DECOMMIT)
+
+static void commit_error(void *p)
+{
+  abort();
+}
+
+#define RESERVE_AND_COMMIT_PAGES(n)                                            \
+  LJ_WIN_VALLOC(0, n, MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE)
+
+#endif
+
+static SRWLOCK g_lock = SRWLOCK_INIT;
+#define LOCK() AcquireSRWLockExclusive(&g_lock)
+#define UNLOCK() ReleaseSRWLockExclusive(&g_lock);
+
+#else
+#error "No page allocation OS support!"
+#endif
+
+#define RESERVE_SIZE 64 * 1024 * 1024
+/* output >= input */
+#define PTRALIGN_UP(x, y)                                                      \
+  (void *)(((uintptr_t)(x) + ((y)-1)) & ~(uintptr_t)((y)-1))
+/* output <= input */
+#define PTRALIGN_DOWN(x, y) (void *)(((uintptr_t)(x)) & ~(uintptr_t)((y)-1))
+
+struct arena_global {
+  void **freelist;
+  uint32_t freelist_sz;
+  uint32_t freelist_at;
+
+  void **ptrs;
+  uint32_t ptrs_sz;
+  uint32_t ptrs_at;
+
+  uint8_t *at;
+  uint8_t *end;
+
+  void *hint;
+
+  uint32_t refs;
+} static g_arena;
```

Comment:
Is there a reason why this `g_arena` global isn't being tucked under `global_State`?

Reply:
Well that one is shared among all states on the theory that the maximum consumption for all states unified is less than the the maximum for all states separately. But it's easy enough to do one per state instead. This is the replaceable allocator

Reply:
> the theory that the maximum consumption for all states unified is less than the the maximum for all states separately

The main problem isn't maximum consumption, it's threading. LuaJIT works when multi-threaded; while each state can only run on one thread, multiple states can run on multiple threads. By introducing a global mutex and global variable into the equation, you're preventing LuaJIT from being able to run multi-threaded at any meaningful volume.

It also adds a minor concern of potential crosstalk or resource hogging issues. Plus linking compatibility, which mike is huge on as it makes LuaJIT a true "drop in" replacement. (I'm not sure if the changes you made actually introduced any dependencies.)

> But it's easy enough to do one per [global] state instead.

This would be ideal in my mind. Drop the locking logic, and pass the arena state around as userdata. You've already completed the heavy lifting for doing that technique, anyways.

Thanks!

Reply:
At 16 threads there is basically zero difference between the lock and separate state versions on my machine, hammering allocation faster than the VM could. Contention on the lock is going to be fairly low given one call represents ~8000 tables + small arrays, and in any case the OS will hold an exclusive lock around the entire mmap call, plus probably a reader on that same lock around each page fault. You'd see effects if you went maybe 50 threads wide but there's little reason to do that in a single process

But since I have written it now to test it'll get wrapped into the next cleanup push